### PR TITLE
flatten_indices: remove syncs

### PR DESCRIPTION
**Pull request by nikitaved**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95165
* #94991
* #94596
* #94406
* __->__ #94401

### CODE REVIEW

**Review by cpuhrsch — APPROVED** [review]

File: aten/src/ATen/native/sparse/FlattenIndicesCommon.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      .build();
+
+    {
+      const auto* RESTRICT ptr_indices = indices.data_ptr<index_t>();
+
+      KernelLauncher<kernel_t>::launch(iter,
+          // NOTE: capture by value required by CUDA
+          [=] FUNCAPI (int64_t nnz_idx) -> int64_t {
+          const auto* RESTRICT ptr_indices_dim = ptr_indices + nnz_idx * indices_nnz_stride;
+          auto hash = static_cast<int64_t>(0);
```
Comment (pearu) [inline aten/src/ATen/native/sparse/FlattenIndicesCommon.h:87]:
```suggestion
          int64_t hash = 0;
```

**Review by pearu — APPROVED** [review]
LGMT! Thanks, @nikitaved!

**Comment by malfet** [comment]
This causes internal link failures, trying to investigate:
```
stderr: ld.lld: error: undefined symbol: at::native::DispatchStub<at::Tensor (*)(at::Tensor const&, c10::ArrayRef<long long>), at::native::flatten_indices_stub>::DEFAULT
>>> referenced by DispatchStub.h:137 (buck-out/gen/fe3a39b8/xplat/caffe2/aten_headerAndroid#header-mode-symlink-tree-with-header-map,headers/ATen/native/DispatchStub.h:137)
>>>               buck-out/gen/fe3a39b8/xplat/caffe2/aten_cpuAndroid#android-x86,compile-pic-SparseTensorUtils.cpp.oeb82483d/aten/src/ATen/SparseTensorUtils.cpp.o:(at::sparse::flatten_indices(at::Tensor const&, c10::ArrayRef<long long>, bool))
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Pull request closed without merging**